### ISSUE TITLE
bug: reduced timeouts on API breaks checks on slow clusters

### ISSUE THREAD

**Issue by radu-matei** [linkerd/linkerd2#2011]
related: #553, #586 

Hey!
While I clearly understand the motivation behind the 5 seconds timeout on checks, the fact that this is not configurable breaks the following use case - I have a dev cluster made of burstable instances - this means that most of the time, they are really, really slow, which makes the Kubernetes API server respond in most cases in a little over 5 seconds (and because it is a cheap dev cluster, I suspect there are other people using them in the same way).

Now, because of the change in #586  (and because the `--wait` flag for `linkerd check` is rather unclear in functionality - `Retry and wait for some checks`), it means that there is no way for me to successfully `linkerd check --pre` on this type of clusters.

One solution would be to add a new flag for this, or re-use the existing `--wait` flag (although I'd avoid changing the behaviour of a flag) - in any case, the default behaviour would still be dictated by #586 , thus not affecting the majority of clusters where response times are much lower.

I'd be more than glad to open a PR for this, but before I do that, I'd like an opinion first.

Thanks!

Edit: as I'm playing around the repo, it feels like the timeout for the Kuberentes API should be a global flag, as it also impacts installs and other configs.

**Comment by klingerf** [comment]
@radu-matei Thanks for bringing this to our attention, and for all of the additional info about your setup. I agree that the timeout should be configurable via a global command line flag, and I'd be more than happy to review that change.

For the `check` command specifically, my preference is to use a separate flag instead of the `--wait` flag, since the `--wait` timeout spans all requests made in the process of completing a check run, whereas this timeout would apply to individual requests. Maybe we could call it `--request-timeout` and default it to 5 seconds?

Unfortunately, the 5-second timeout is hardcoded throughout `pkg/...`. It would be great if we could consolidate all of these as part of the change:

```
./pkg/k8s/api.go:	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
./pkg/k8s/api.go:	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
./pkg/k8s/api.go:	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
./pkg/healthcheck/healthcheck.go:			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
./pkg/version/version.go:	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
./pkg/version/version.go:	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

And big 👍 to improving the help text for the `--wait` and `--request-timeout` flags to make it more clear how they impact CLI behavior.

Note that this change also relates to #1854, which involves the pre-checks that we run prior to querying the public API from the CLI.

**Comment by radu-matei** [comment]
One point I'd like to raise before starting a fix for this - does the current version of Linkerd ensure there are non-breaking API changes?

The fix would most likely involve passing a `timeout` parameter to the public methods for the packages mentioned above (which changes the method signature, which is a breaking change) - if we ensure non-breaking API changes, then we would need to create new methods - for example, for `GetVersionInfo`, we would also have `GetVersionInfoWithTimeout` (and so on).

Otherwise, we'll just add the parameter to the methods and fix all method calls.

**Comment by klingerf** [comment]
Good question! And no, at this point we make no guarantees on the stability of publicly exported methods for any of the packages in the project. It should be no problem to change those method signatures.

But in the interest of compatibility, maybe we should update those methods to accept a full context object with the timeout already applied, rather than accepting a timeout param and creating a new context inside each method? That would hopefully lead to more stable method signatures going forward.

**Comment by leonfs** [comment]
We've been having some issues with the timeout as well. We have an AKS (Azure) cluster that doesn't shine for its speed. When trying to install linkerd, we had checks failing too. 
```
kubernetes-api: can query the Kubernetes API...............................[FAIL] -- Get https://mycluster.hcp.uksouth.azmk8s.io:443/version: context deadline exceeded
```
Running the linkerd client within a bastion host, the checks passed. 

It would be great to have a timeout property.

**Comment by alpeb** [comment]
We released Edge-19.2.1 yesterday, which has much larger timeouts for these checks. Please give it a try and let us know how it goes:
https://github.com/linkerd/linkerd2/releases/tag/edge-19.2.1

**Comment by klingerf** [comment]
The increased timeouts have now been released as part of the stable-2.2.0 release; am going to close this.
